src/compiler/tsbuildPublic.ts

```diff
@@ -1459,80 +1506,182 @@ namespace ts {
                     };
                 }
 
-                // Check oldest output file name only if there is no missing output file name
-                // (a check we will have skipped if this is a forced build)
-                if (!force && !missingOutputFileName) {
-                    // If the upstream project's newest file is older than our oldest output, we
-                    // can't be out of date because of it
-                    if (refStatus.newestInputFileTime && refStatus.newestInputFileTime <= oldestOutputFileTime) {
-                        continue;
-                    }
+                if (!force) (referenceStatuses ||= []).push({ ref, refStatus, resolvedRefPath, resolvedConfig });
+            }
+        }
+        if (force) return { type: UpToDateStatusType.ForceBuild };
 
-                    // If the upstream project has only change .d.ts files, and we've built
-                    // *after* those files, then we're "psuedo up to date" and eligible for a fast rebuild
-                    if (refStatus.newestDeclarationFileContentChangedTime && refStatus.newestDeclarationFileContentChangedTime <= oldestOutputFileTime) {
-                        pseudoUpToDate = true;
-                        upstreamChangedProject = ref.path;
-                        continue;
-                    }
+        // Check buildinfo first
+        const { host } = state;
+        const buildInfoPath = getTsBuildInfoEmitOutputFilePath(project.options);
+        let oldestOutputFileName = "(none)";
+        let oldestOutputFileTime = maximumDate;
+        let buildInfoTime: Date | undefined;
+        let newestDeclarationFileContentChangedTime;
+        if (buildInfoPath) {
+            const buildInfoCacheEntry = getBuildInfoCacheEntry(state, buildInfoPath, resolvedPath);
+            buildInfoTime = buildInfoCacheEntry?.modifiedTime || ts.getModifiedTime(host, buildInfoPath);
+            if (buildInfoTime === missingFileModifiedTime) {
+                if (!buildInfoCacheEntry) {
+                    state.buildInfoCache.set(resolvedPath, {
+                        path: toPath(state, buildInfoPath),
+                        buildInfo: false,
+                        modifiedTime: buildInfoTime
+                    });
+                }
+                return {
+                    type: UpToDateStatusType.OutputMissing,
+                    missingOutputFileName: buildInfoPath
+                };
+            }
+
+            const buildInfo = Debug.checkDefined(getBuildInfo(state, buildInfoPath, resolvedPath, buildInfoTime));
+            if ((buildInfo.bundle || buildInfo.program) && buildInfo.version !== version) {

```

Comment:
Previously we had cache to do this check only once because we dint want to reparse the version all the time, but since its cached we can do it all the time now.. Just saves us maintaining that cache